File: Back-End/src/services/degreeRequirementService.js
const { query } = require('../db');

const mapRequirement = (r) => ({
  requirement_id: Number(r.requirement_id),
  major_id: r.major_id == null ? null : Number(r.major_id),
  major_name: r.major_name || null,
  course_id: r.course_id == null ? null : Number(r.course_id),
  course_code: r.course_code || null,
  course_name: r.course_name || null,
  requirement_type: r.requirement_type,
  is_required: r.is_required,
  credits: r.credits == null ? null : Number(r.credits),
});

const listRequirements = async ({ major_id, course_id, requirement_type } = {}) => {
  const params = [];
  const where = [];
  if (major_id) {
    params.push(Number(major_id));
    where.push(`dr.major_id = $${params.length}`);
  }
  if (course_id) {
    params.push(Number(course_id));
    where.push(`dr.course_id = $${params.length}`);
  }
  if (requirement_type) {
    params.push(String(requirement_type).toLowerCase());
    where.push(`LOWER(dr.requirement_type) = $${params.length}`);
  }

  const result = await query(
    `SELECT dr.requirement_id,dr.major_id,dr.course_id,dr.requirement_type,dr.is_required,dr.credits,
            m.major_name,c.course_code,c.course_name
     FROM degree_requirements dr
     LEFT JOIN majors m ON m.major_id = dr.major_id
     LEFT JOIN courses c ON c.course_id = dr.course_id
     ${where.length ? `WHERE ${where.join(' AND ')}` : ''}
     ORDER BY dr.requirement_id DESC`,
    params
  );

  return result.rows.map(mapRequirement);
};

const getRequirementById = async (id) => {
  const result = await query(
    `SELECT dr.requirement_id,dr.major_id,dr.course_id,dr.requirement_type,dr.is_required,dr.credits,
            m.major_name,c.course_code,c.course_name
     FROM degree_requirements dr
     LEFT JOIN majors m ON m.major_id = dr.major_id
     LEFT JOIN courses c ON c.course_id = dr.course_id
     WHERE dr.requirement_id = $1`,
    [Number(id)]
  );
  return result.rowCount ? mapRequirement(result.rows[0]) : null;
};

const createRequirement = async (payload) => {
  const result = await query(
    `INSERT INTO degree_requirements(major_id,course_id,requirement_type,is_required,credits)
     VALUES($1,$2,$3,$4,$5)
     RETURNING requirement_id,major_id,course_id,requirement_type,is_required,credits`,
    [
      payload.major_id ? Number(payload.major_id) : null,
      payload.course_id ? Number(payload.course_id) : null,
      payload.requirement_type || null,
      payload.is_required ?? true,
      payload.credits == null ? null : Number(payload.credits),
    ]
  );
  return getRequirementById(result.rows[0].requirement_id);
};

const updateRequirement = async (id, payload) => {
  const result = await query(
    `UPDATE degree_requirements
     SET major_id=$1,course_id=$2,requirement_type=$3,is_required=$4,credits=$5
     WHERE requirement_id=$6
     RETURNING requirement_id`,
    [
      payload.major_id ? Number(payload.major_id) : null,
      payload.course_id ? Number(payload.course_id) : null,
      payload.requirement_type || null,
      payload.is_required ?? true,
      payload.credits == null ? null : Number(payload.credits),
      Number(id),
    ]
  );
  return result.rowCount ? getRequirementById(id) : null;
};

const deleteRequirement = async (id) => {
  const result = await query('DELETE FROM degree_requirements WHERE requirement_id=$1 RETURNING requirement_id', [Number(id)]);
  return result.rowCount > 0;
};

module.exports = { listRequirements, getRequirementById, createRequirement, updateRequirement, deleteRequirement };

File: Back-End-Python/app/services/degree_requirement_service.py
"""Mirrors Back-End/src/services/degreeRequirementService.js"""

from sqlalchemy import text
from sqlalchemy.orm import Session


def _map_requirement(r) -> dict:
    return {
        "requirement_id": int(r["requirement_id"]),
        "major_id": int(r["major_id"]) if r["major_id"] is not None else None,
        "major_name": r["major_name"] or None,
        "course_id": int(r["course_id"]) if r["course_id"] is not None else None,
        "course_code": r["course_code"] or None,
        "course_name": r["course_name"] or None,
        "requirement_type": r["requirement_type"],
        "is_required": r["is_required"],
        "credits": float(r["credits"]) if r["credits"] is not None else None,
    }


_BASE_SQL = """
SELECT dr.requirement_id, dr.major_id, dr.course_id, dr.requirement_type,
       dr.is_required, dr.credits, m.major_name, c.course_code, c.course_name
FROM degree_requirements dr
LEFT JOIN majors m ON m.major_id = dr.major_id
LEFT JOIN courses c ON c.course_id = dr.course_id
"""


def list_requirements(
    db: Session,
    major_id: int | None = None,
    course_id: int | None = None,
    requirement_type: str | None = None,
) -> list:
    where = []
    params: dict = {}

    if major_id:
        where.append("dr.major_id = :major_id")
        params["major_id"] = int(major_id)
    if course_id:
        where.append("dr.course_id = :course_id")
        params["course_id"] = int(course_id)
    if requirement_type:
        where.append("LOWER(dr.requirement_type) = :requirement_type")
        params["requirement_type"] = str(requirement_type).lower()

    where_sql = ("WHERE " + " AND ".join(where)) if where else ""
    rows = db.execute(
        text(f"{_BASE_SQL} {where_sql} ORDER BY dr.requirement_id DESC"), params
    ).mappings().all()
    return [_map_requirement(r) for r in rows]


def get_requirement_by_id(db: Session, requirement_id: int) -> dict | None:
    row = db.execute(
        text(f"{_BASE_SQL} WHERE dr.requirement_id = :rid"),
        {"rid": int(requirement_id)},
    ).mappings().first()
    return _map_requirement(row) if row else None


def create_requirement(db: Session, payload: dict) -> dict:
    db.execute(
        text(
            "INSERT INTO degree_requirements(major_id, course_id, requirement_type, is_required, credits) "
            "VALUES (:mid, :cid, :rtype, :req, :credits)"
        ),
        {
            "mid": int(payload["major_id"]) if payload.get("major_id") else None,
            "cid": int(payload["course_id"]) if payload.get("course_id") else None,
            "rtype": payload.get("requirement_type"),
            "req": bool(payload.get("is_required", True)),
            "credits": float(payload["credits"]) if payload.get("credits") is not None else None,
        },
    )
    row = db.execute(
        text("SELECT requirement_id FROM degree_requirements ORDER BY requirement_id DESC LIMIT 1")
    ).mappings().first()
    db.commit()
    return get_requirement_by_id(db, int(row["requirement_id"]))


def update_requirement(db: Session, requirement_id: int, payload: dict) -> dict | None:
    db.execute(
        text(
            "UPDATE degree_requirements SET major_id=:mid, course_id=:cid, "
            "requirement_type=:rtype, is_required=:req, credits=:credits "
            "WHERE requirement_id=:rid"
        ),
        {
            "mid": int(payload["major_id"]) if payload.get("major_id") else None,
            "cid": int(payload["course_id"]) if payload.get("course_id") else None,
            "rtype": payload.get("requirement_type"),
            "req": bool(payload.get("is_required", True)),
            "credits": float(payload["credits"]) if payload.get("credits") is not None else None,
            "rid": int(requirement_id),
        },
    )
    db.commit()
    return get_requirement_by_id(db, requirement_id)


def delete_requirement(db: Session, requirement_id: int) -> bool:
    result = db.execute(
        text("DELETE FROM degree_requirements WHERE requirement_id=:rid RETURNING requirement_id"),
        {"rid": int(requirement_id)},
    ).mappings().first()
    db.commit()
    return result is not None
